Clean-up the use of RFC2119 keywords (#4550)
knative/serving: docs/runtime-contract.md

([v1.0.1](https://github.com/opencontainers/runtime-spec/blob/v1.0.1/spec.md))
is the basis for this document. When this document and the OCI specification
conflict, this document is assumed to override the general OCI recommendations.
Where this document does not specify behavior, runtime implementations SHOULD

> be OCI compliant with respect to those features.  Additionally, the core Knative definition

Reviewer: Thanks, this is much clearer.

assumes the
[Linux Container Configuration](https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md).

In particular, the default Knative implementation relies on Kubernetes behavior
to implement container operation. In some cases, current Kubernetes behavior in

> 2018 is not as performant as specified in this documentation. The goal of the

Reviewer: I don't think this is "recommended", I think this is "envisioned".

Author: fixed

Knative authors is to push as much of the needed functionality into Kubernetes

> and/or Istio as possible, rather than implementing reach-around layers.

Reviewer: Here's a line which hasn't aged well. How about:

```suggestion
and/or HTTP routers as possible, rather than implementing reach-around layers.
```

Reviewer: You should probably make this edit yourself rather than relying on Github suggestion, since that will anger the mighty CLA-bot.

Author: done

> 

Reviewer 2: I cannot comment on it here, but lines 73-74 state:

"containers which are operated as part of a serverless platform SHOULD observe the following properties:"

The use of the keyword here should be replaced with a non keyword.

Author: yep good catch - fixed

### State

The general OCI state might not be available for introspection within the

> container, and might only be visible to the system operator or platform provider.

Reviewer: I think `MAY` might be correct here, but I think @dgerd has thought about this more recently than I.

Reviewer 2: I would actually prefer removing this as it:

1. Seems duplicative with https://github.com/knative/serving/blob/master/docs/runtime-contract.md#operations
2. Provides little guidance/value to users or platform providers

Author: Since I thought it was non-normative anyway, I'm ok with removing it - done!

  contents from a particular execution. Because containers (particularly failing
  containers) can experience frequent starts, operators or platform providers
  SHOULD limit the total space consumed by these failures.

> - A container SHOULD write its own termination message to `/dev/termination-log`

Reviewer 2: I would make this 'ought'.

Personally I would like to see us remove this in a future change. See discussion here: https://github.com/knative/serving/pull/4035/files#r283528430

Author: yep - we have no say over this. Fixed.

  by default. If no message is written by the container, the last few lines of

>   log output SHOULD be reported as the execution error (i.e. by

Reviewer 2: We don't do this today and I don't think we plan to do this. K8s exposes this on the Pod which is not part of the Knative API and is difficult to aggregate up to the Revision.

Author: hmm, who is supposed to "report this error" ? Knative? the OCI runtime?

Reviewer 2: I have read it in the past as Knative. A Knative user does not have interaction with the OCI runtime. A user is able to see what is visible through our API, monitoring/logging extension points, anything observable within the container, and anything observable about the lifecycle of the container.

If this is only visible in the OCI API then it is not visible to the end user and is thus could be omitted from the runtime without consequence to user experience.

Reviewer: See the discussion in #4035 -- it turns out that this information is collected at the Pod level, but is not aggregated beyond that level.

Having looked into it further, I'm concerned about the cascading update scenario that aggregating this to status information in the apiserver above the Pod level enables. Consider 1000 containers which crash in a coordinated manner -- that's a lot of extra updates (and possibly conflicts which need to be retried) that put load on the apiserver at exactly the wrong time.

Reviewer: I think the discussion in #4035 is leaning towards removing this text altogether -- there's no place in the API for these errors to be surfaced beyond the Pod level, and Pods are really ephemeral, so the termination message isn't so helpful.

Author: I'm trying to limit the semantic changes made to this PR but I'll remove this text anyway since you guys are convinced it's wrong. I removed the entire bullet since if I remove the last sentence then there's no normative text left - so there's no point in keeping the rest of it.

collected and retained in a developer-accessible logging repository.
(TODO:[docs#902](https://github.com/knative/docs/issues/902)).

> Within the container, pipes and file descriptors MAY be used to communicate

Reviewer 2: The MAY keyword doesn't seem appropriate here. If a container MAY be able to do this it implies that the platform MUST allow pipes and file descriptors for IPC. Using MUST and phrasing from the platform point-of-view seems like the more appropriate wording, but the statement seems like overkill.

I cannot think of a container runtime that doesn't allow this and I cannot think of a good reason for a future runtime to disable this.

@evankanderson Any thoughts on why this is here?

Author: changing it to a "can" since we have no say over what happens within the container w.r.t. these pipes - IOW, I think this statement is out of scope of Knative.

Reviewer 2: If we are going to change to "can" I would prefer removal of it. Keeping this document concise keeps it readable and usable. I don't think it clarifies much for platform providers or users as a "can" statement.

Reviewer: I think this was added at one point due to an internal discussion about required gvisor capabilities. In any case, it was a point-in-time thing, and I agree with removing it, unless we're going to spell out a complete set of linux syscalls which are in-scope.

- `h2c`: HTTP/2 transport, as described in
  [section 3.4 of the HTTP2 spec (Starting HTTP/2 with Prior Knowledge)](https://http2.github.io/http2-spec/#known-http)

Developers ought to use automatic content negotiation where available,
and MUST NOT set the `name` field to arbitrary values, as additional transports

> might be defined in the future. Developers can assume all traffic is

Reviewer: Why did we remove MUST here? I think #4035 moves these to a "advice to the developer" section, but the requirement here is intended to include:

1. Don't assume that the other end of the TCP connection is the client
2. Don't need to implement SSL in the container

Author: I removed the MUST because we can't really control what the developers "assume". If we want to add something normative here then it needs to be a constraint on the Kn impl - so something like: `Implementations MUST send all traffic through L7 proxies`.  Do you want that instead?

Reviewer 2: I am fine with this change. 

This runtime contract is still pretty long and I don't expect developers to read through it. The language is meant for operators and contributors.

I would eventually like to move these suggestions to a shorter user-focused document. I intend to break this file out into its own PR ( https://github.com/knative/serving/pull/4035/files#diff-2b3ccedac0c1b327ad7d674942f6cf56 ).

- `h2c`: HTTP/2 transport, as described in
  [section 3.4 of the HTTP2 spec (Starting HTTP/2 with Prior Knowledge)](https://http2.github.io/http2-spec/#known-http)

> Developers SHOULD use automatic content negotiation where available,

Reviewer 2: 'ought' seems appropriate here. 

Also just FYI that we currently do not have automatic content negotiation. See https://github.com/knative/serving/pull/4286

Author: used "ought"

> and MUST NOT set the `name` field to arbitrary values, as additional transports

Author: @dgerd is the "MUST NOT" here overstepping too? Seems like it given we're removing the other keywords from this paragraph.

Reviewer 2: The "MUST NOT" is important to allow us to reserve the 'name' field for future usage. It could be argued that this is more of an API constraint than a runtime constraint, but it is important and I am okay with duplicating it here for clarity.

For example, if HTTP/3.0 comes along and we want to allow people to specify "http3" we can only enable this we know that no existing Knative Configurations allow "http3"; otherwise this will be a breaking change.

I am okay rephrasing this if you think it is unclear as is.

Reviewer: I think the "MUST NOT" suggests that validation during admission control is allowed here, and containers with other port names may be rejected.

Author: ok I'll leave it as-is.

> might be defined in the future. Developers MUST assume all traffic is

Reviewer 2: The "Developers MUST assume" and "Developers MUST NOT assume" should use non-keywords.

Author: done

- `initialDelaySeconds` set to 0
- `periodSeconds` set to platform-specific value

> Setting `initialDelaySeconds` to a value of 0 will enable scaling in response

Reviewer 2: I like the removal of SHOULD NOT. Two nits:

(1) While line 275 explains that 0 is the default this line seems to encourage explicitly setting to 0.
(2) This explains what happens when it is set to 0, but does not explain what happens when it goes past zero or why a developer should prefer to keep it unset/0.

Here is what I suggest:
```
Setting `initialDelaySeconds` to a value greater than 0 impacts container startup time (aka cold start time) as a container will not serve traffic until the probe succeeds.
```

Reviewer: Agree with @dgerd but s/will not serve traffic until the probe succeeds/cannot receive traffic until the probe has started/, I think.

Author: Done

- `periodSeconds` set to platform-specific value

> In order to enable scaling in response to load, developers SHOULD NOT set the

Reviewer 2: "Developers SHOULD NOT" should be changed to a non-keyword. We allow setting initialDelaySeconds greater than zero and developers might do this.

"SHOULD aim to minimize" should be changed to a non-keyword.

Author: see what you think of the new wording

- `initialDelaySeconds` set to 0
- `periodSeconds` set to platform-specific value

In order to enable scaling in response to load, setting `initialDelaySeconds`

> to a value greater than 0 can be used, while striving to minimize container

Reviewer: I think this sentence now reads backwards -- it suggests that `initialDelaySeconds > 0` may improve scaling response, which is almost certainly not true.

Author: oops, yeah I got that backwards.... let me fix that

### Default Filesystems

> The OCI specification describes a default container environment which might be

Reviewer: ```suggestion
The OCI specification describes a default container environment which can be
```

I think "can" reads more clearly here.

Author: done

used for many different purposes, including containerization of existing legacy
or stateful processes which might store substantial amounts of on-disk state. In a
scaled-out, stateless environment, container startup and teardown is accelerated
when on-disk resources are kept to a minimum. Additionally, developers might not
have access to the container filesystems (or the containers might be rapidly
recycled), so log aggregation SHOULD be provided.

> In addition to the filesystems suggested in the OCI, the following filesystems

Reviewer: I think this is RECOMMENDED if it's coming from OCI.

Reviewer 2: +1

Author: OCI might recommend them, but I'm not sure the use of the word here is normative since we're not actually adding a new constraint we're just saying "they tell us to do this".  But let me think about it....

Reviewer 2: IMO it would be great to add the OCI recommended filesystems inline with a link to where they come from. No one is going to click out to find what are the recommended filesystems.

Reviewer: Ha! The OCI spec is actually SHOULD:

https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#default-filesystems

### Default Filesystems

The OCI specification describes a default container environment which can be
used for many different purposes, including containerization of existing legacy
or stateful processes which might store substantial amounts of on-disk state. In a
scaled-out, stateless environment, container startup and teardown is accelerated
when on-disk resources are kept to a minimum. Additionally, developers might not

> have access to the container filesystems (or the containers might be rapidly

Reviewer: I just noticed that this sentence doesn't make sense with "container filesystems". I think it was meant to say "durable storage" or "durable filesystems" (i.e. it would be fine to put all the container writes into tmpfs and throw them away when the container exits).

Author: 'm reading this as trying to say that the developer might not be able to see the container's filesystem so they can't look at log files, etc... I think the issue might be that it should be "container's" not "container".  WDYT?

Reviewer 2: I think the `container` -> `container's` change is sufficient here, but it could be updated further for clarity.

My understanding here is that: (1) Developers don't necessarily have access to the underlying host so you can't go looking at the stopped container contents locally. (2) Unlike docker which relies on a user to `rm` a stopped container or run `prune`, containers in this environment are likely to be cleaned up quickly after exiting.

Author: added the apostrophe

| `/var/log` | MUST be a directory with write permissions for logs storage. Implementations MAY permit the creation of additional subdirectories and log rotation and renaming. |
| `/dev/log` | MUST be a writable socket to syslog                                                                                                                              |

> In addition, the following files MAY be overridden by the runtime environment to

Reviewer 2: `/etc/resolve.conf` is listed as "SHOULD be set". It is somewhat confusing to have the MAY paragraph prefix this table. 

I would opt to change the word 'may' in the leading paragraph and rely on the Description to specify the keywords.

Author: I was a bit bothered too as I was tweaking this... I'll do that.

| `/tmp`     | MUST be Read-write.<p>SHOULD be backed by tmpfs if disk load is a concern.                                                                                       |
| `/var/log` | MUST be a directory with write permissions for logs storage. Implementations MAY permit the creation of additional subdirectories and log rotation and renaming. |

In addition, the following constraints are specified for the overridden files

> indicated:

Reviewer: This rewrite removes the DNS rationale. What about:

"To enable DNS resolution, the following files might be overwritten at runtime:"

Author: much better - done

permitted to the processes running inside the container environment. A seccomp
sandbox MAY be enforced by the platform operator; any such application profiles
SHOULD be configured and applied in a consistent mechanism outside of the

> container specification. As the seccomp policy might be part of the platform

Reviewer 2: Why not 'MAY'?

Author: I may have read it wrong, but I didn't think the "may" here was our way of saying "you can do this if you want" but rather it was trying to explain that something this thing happens and we're not really making a statement as to whether it's ok or not.  Was it intended to introduce a constraint or was it just explaining the state of the world?

Reviewer: I think we want to say that it is okay; sharing a kernel is sufficiently exciting from a security perspective that we should be clear that enabling defense mechanisms here is prudent (though not required -- something like Kata Containers or gvisor has different mechanisms for protection).

Author: if we want MAY in there, then I think we need to remove the "As the..." part, because then it reads kind of weird as a normative statement. See what you think of the new text

permitted to the processes running inside the container environment. A seccomp
sandbox MAY be enforced by the platform operator; any such application profiles
SHOULD be configured and applied in a consistent mechanism outside of the

> container specification. Seccomp policy MAY be part of the platform

Reviewer 2: ```suggestion
container specification. A Seccomp policy MAY be part of the platform
```

Author: done

> security configuration so that operators can tune this over time as the

Reviewer 2: ```suggestion
security configuration that operators can tune over time as the
```

Author: done

> article in the kernel documentation has more information about mount
> propagation.

> This option SHOULD only be set by the operator or platform provider, and MUST

Reviewer 2: ```suggestion
This option MAY be set by the operator or platform provider, and MUST
```

The second part of the sentence clarifies the 'only' part so that is unnecessary. We don't provide any real guidance or requirement here so SHOULD seems to be too strong of a word choice.

Author: ok done

NOT be configurable by the developer. As mount propagation might be part of the

> platform security hardening, operators might tune this over time as the threat

Reviewer: If 474 is MAY, this should be MAY as well.

Reviewer 2: +1 I would make both of these MAY

Author: Are we really adding constraints here or just stating how things are? It didn't come across to me like we were being normative. For example, the phrase "person X might do something over time" comes across to me more like "guidance" or explaining what's going on in the system rather than saying "make sure person X is given the option of doing ....".  If we really are saying "make sure the operation is given the option of doing this" then I think we need to word it more strongly so it's not so ad hoc-ish.

Reviewer: I think we want to explicitly word it so that these tools are available for operators.

Would "Operators MAY disable or force this setting to a particular value" be better?

Author: see if the new text is ok

environment changes.

### Masked Paths

> This option SHOULD only be set by the operator or platform provider, and MUST NOT

Reviewer 2: ```suggestion
This option MAY be set by the operator or platform provider, and MUST NOT
```

Author: done

be configurable by the developer. As masked paths might be part of the platform

> security hardening, operators might tune this from time to time as the threat

Reviewer: Ditto.

Author: Came concern as above - is this guidance/explanatory text or constraints?

Author: added new text to align with the other 2 things like this. SWYT

> article in the kernel documentation has more information about mount
> propagation.

This option MAY be set by the operator or platform provider, and MUST
NOT be configurable by the developer. Mount propagation MAY be part of the

> platform security configuration so that operators can tune this over time

Reviewer 2: ```suggestion
platform security configuration that operators can tune over time
```

Author: done

as the threat environment changes.

### Masked Paths

This option MAY be set by the operator or platform provider, and MUST NOT
be configurable by the developer. Masked paths MAY be part of the platform

> security configuration so operators can tune this from time to time as the

Reviewer 2: ```suggestion
security configuration that operators can tune over time as the
```

Author: done
